src/meta/app/src/tenant_key.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::hash::Hash;
+use std::hash::Hasher;
+
+use databend_common_meta_kvapi::kvapi;
+
+use crate::tenant::Tenant;
+
+/// Defines the in-meta-service data for some resource that belongs to a tenant.
+/// Such as `PasswordPolicy`.
+///
+/// It includes a prefix to store the `ValueType`.
+/// This trait is used to define a concrete [`TIdent`] can be used as a `kvapi::Key`.
+pub trait TenantResource {
+    const PREFIX: &'static str;
+    type ValueType: kvapi::Value;
+}
+
+/// `[T]enant[Ident]` is a common meta-service key structure in form of `<PREFIX>/<TENANT>/<NAME>`.
```

Comment:
Coooooool. I love this comment. If not, maybe serval years, I will think it's [T]emplate[Ident]